MCR-3096 prepare for ocfl

Mewel opened the pull request:
-Support different digest algorithms like SHA-512
-Extend NIOFS API
-Support Versioning Paths
-Improve Transaction Handling

[Link to jira](https://mycore.atlassian.net/browse/MCR-3096).

Mewel committed:
MCR-3096 prepare for ocfl

-Support different digest algorithms like SHA-512
-Extend NIOFS API
-Support Versioning Paths
-Improve Transaction Handling

Mewel committed:
MCR-3096 organize imports

Mewel committed:
fix some codacy bugs

Mewel committed:
MCR-3096 fix codacy again

Mewel committed:
MCR-3096 only restart database transaction

Mewel committed:
MCR-3096 add equals/hashCode for MCRDigest

Mewel committed:
MCR-3096 improve javadoc & add ReadOnlyException

sebhofmann commented on mycore-base/src/main/java/org/mycore/common/digest/MCRDigest.java:
```diff
@@ -0,0 +1,111 @@
+ * @see MCRMD5Digest
+ * @see MCRSHA512Digest
+ */
+public abstract sealed class MCRDigest permits MCRMD5Digest, MCRSHA512Digest {
```
I don't see any reason why MCRDigest should not allow other Implementations.

sebhofmann commented on mycore-base/src/main/java/org/mycore/common/digest/MCRMD5Digest.java:
```diff
@@ -0,0 +1,52 @@
+
+    @Override
+    protected void validate() throws MCRDigestValidationException {
+        if (this.value.length() != 32 || !MD5_HEX_PATTERN.matcher(value).matches()) {
```
Better use HexFormat and parseHex.

sebhofmann commented on mycore-base/src/main/java/org/mycore/common/digest/MCRDigest.java:
```diff
@@ -0,0 +1,111 @@
+
+    }
+
+    protected String value;
```
A digest should be bytes and there should be a toHexString method.

Mewel committed:
Merge branch 'main' into issues/MCR-3096-prepare-for-ocfl

sebhofmann commented on mycore-base/src/main/java/org/mycore/datamodel/niofs/MCRAbstractFileSystem.java:
```diff
@@ -144,14 +146,14 @@ public boolean verifies(MCRPath path, MCRFileAttributes<?> attrs) throws NoSuchF
-        boolean returns = md5Sum.matches(attrs.md5sum());
+        boolean returns = digest.matches(attrs.digest().getValue());
         if (!returns) {
             LogManager.getLogger(getClass()).warn("MD5sum does not match: {}", path);
```
Should use digest algorithm here and not asume it is md5 (in the log line)

sebhofmann commented on mycore-base/src/main/java/org/mycore/datamodel/niofs/MCRAbstractFileSystemProvider.java:
```diff
@@ -0,0 +1,178 @@
+        return getPath(pathInfo.owner(), pathInfo.path(), getFileSystem());
+    }
+
+    public MCRPath getPath(String owner, String path, MCRAbstractFileSystem fs) {
```
Is this really required? It can default to getFileSystem()

Mewel replied:
@yagee-de Can you elaborate on this thomas? Should we add a FileSystem member variable for MCRPath to avoid the anonymous class?

yagee-de replied:
MCRAbstractFileSystemProvider already has a method "getFileSystem()" It should be used instead of a parameter in getPath(). Makes little sense to construct an MCRPath in FileSystem "A" with the Implementation of FileSystem "B".

sebhofmann commented on mycore-base/src/main/java/org/mycore/datamodel/niofs/MCRBasicFileAttributeViewProperties.java:
```diff
@@ -0,0 +1,115 @@
+
+    private static final String IS_OTHER_NAME = "isOther";
+
+    public static final HashSet<String> ALLOWED_ATTRIBUTES = Sets.newHashSet(ALL, SIZE_NAME, CREATION_TIME_NAME,
```
Should be immutable.

sebhofmann commented on mycore-base/src/main/java/org/mycore/datamodel/niofs/MCRBasicFileAttributeViewProperties.java:
```diff
@@ -0,0 +1,115 @@
+
+    private static final String ALL = "*";
+
+    private static final String SIZE_NAME = "size";
```
The attributes should be an enum.

sebhofmann commented on mycore-base/src/main/java/org/mycore/datamodel/niofs/MCRBasicFileAttributesProxy.java:
```diff
@@ -0,0 +1,98 @@
+
+    protected MCRFileAttributes<T> proxy() {
+        if (this.proxy == null) {
+            synchronized (lock) {
```
The synchronized should use this.proxy and lock should be removed.

Mewel replied:
Would lead to: synchronization on a non-final field 'this.proxy'

sebhofmann commented on mycore-base/src/main/java/org/mycore/datamodel/niofs/MCRDefaultFileAttributes.java:
```diff
@@ -0,0 +1,147 @@
+    }
+
+    /* (non-Javadoc)
+     * @see java.nio.file.attribute.BasicFileAttributes#creationTime()
```
I think its better to use @inheritDoc

sebhofmann commented on mycore-base/src/main/java/org/mycore/common/digest/MCRDigest.java:
```diff
@@ -0,0 +1,111 @@
+    /**
+     * Enumeration of supported digest algorithms.
+     */
+    public enum Algorithm {
```
This enum is okay to avoid magic strings, but i would not make it a requirement for MCRDigest.

sebhofmann commented on mycore-base/src/main/java/org/mycore/datamodel/niofs/MCRVersionedFileSystemProvider.java:
```diff
@@ -0,0 +1,232 @@
+
+    @Override
+    public boolean isHidden(Path path) {
+        MCRVersionedPath.toVersionedPath(path);
```
Is this required?

sebhofmann commented on mycore-restapi/src/main/java/org/mycore/restapi/v1/utils/MCRJSONFileVisitor.java:
```diff
@@ -104,7 +104,7 @@ private void addBasicAttributes(Path path, BasicFileAttributes attrs) throws IOE
             jw.name("contentType").value(MCRContentTypes.probeContentType(path));
             if (attrs instanceof MCRFileAttributes<?> fileAttrs) {
-                jw.name("md5").value(fileAttrs.md5sum());
+                jw.name("md5").value(fileAttrs.digest().getValue());
```
I think its ok to deliver a sha512 hash as md5, because the api would be more stable for old applications. Alternative would be using the digest algo name as key.

sebhofmann commented on mycore-restapi/src/main/java/org/mycore/restapi/v2/MCREventHandler.java:
```diff
@@ -287,7 +287,7 @@ public void sendEvent(MCREvent evt, Path path, BasicFileAttributes attrs) {
             file.addProperty("size", attrs.size());
             file.addProperty("modified", attrs.lastModifiedTime().toInstant().toString());
-            file.addProperty("md5", ((MCRFileAttributes) attrs).md5sum());
+            file.addProperty("md5", ((MCRFileAttributes) attrs).digest().getValue());
```
I think its ok to deliver a sha512 hash as md5, because the api would be more stable for old applications. Alternative would be using the digest algo name as key.

Mewel committed:
MCR-3096 use byte[] array for digest & use class for digest algorithm

Mewel committed:
MCR-3096 fix checkstyle, use enum in FileAttributeViewProperties, use @inheritDoc, add some javadoc

Mewel committed:
MCR-3096 simplify fs implementation

- rm empty path to avoid infinite recursive calls
- add fs to path for easier access

Mewel committed:
MCR-3096 add license information

Mewel committed:
MCR-3096 improve Path API for versioned content

Mewel committed:
MCR-3096 fix codacy bugs

yagee-de commented on mycore-base/src/main/java/org/mycore/common/digest/MCRMD5Digest.java:
```diff
@@ -0,0 +1,62 @@
+        }
+    }
+
+    public static class Md5Algorithm extends Algorithm {
```
naming: should be MD5… not Md5…

yagee-de commented on mycore-base/src/main/java/org/mycore/datamodel/niofs/MCRBasicFileAttributeViewProperties.java:
```diff
@@ -0,0 +1,195 @@
+        ALLOWED_ATTRIBUTES = Arrays.stream(Attribute.values())
+            .map(Attribute::getName)
+            .collect(Collectors.toSet());
+    }
```
initialize when declerating ALLOWED_ATTRIBUTES

Mewel replied:
Any reason why you don't like static blocks?

yagee-de replied:
I <3 static blocks! Use of static blocks is only required when catching an exception is required. Usually it's best to initialize static fields right away.

yagee-de commented on mycore-base/src/main/java/org/mycore/datamodel/niofs/MCRBasicFileAttributeViewProperties.java:
```diff
@@ -0,0 +1,195 @@
+    /**
+     * Enum representing the various file attributes that can be accessed.
+     */
+    private enum Attribute {
+        ALL("*"),
+        SIZE_NAME("size"),
+        CREATION_TIME_NAME("creationTime"),
+        LAST_ACCESS_TIME_NAME("lastAccessTime"),
+        LAST_MODIFIED_TIME_NAME("lastModifiedTime"),
+        FILE_KEY_NAME("fileKey"),
+        IS_DIRECTORY_NAME("isDirectory"),
+        IS_REGULAR_FILE_NAME("isRegularFile"),
+        IS_SYMBOLIC_LINK_NAME("isSymbolicLink"),
+        IS_OTHER_NAME("isOther");
```
Isn't really used as enum anywhere. It's extended in sub classes and always converted to String or from String

Mewel replied:
I just copied this code from the old class. What do you want to have?
```
private static final String ALL = "*";
...
```

yagee-de commented on mycore-base/src/main/java/org/mycore/datamodel/niofs/MCRBasicFileAttributesProxy.java:
```diff
@@ -0,0 +1,98 @@
+        if (this.proxy == null) {
+            synchronized (lock) {
+                if (this.proxy == null) {
+                    this.proxy = loadProxy();
```
if loadProxy() fails every call to it will result in a new request to loadProxy(). So it will not load once but many times and can result in unforeseen exceptions.

yagee-de commented on mycore-base/src/main/java/org/mycore/datamodel/niofs/MCRBasicFileAttributesProxy.java:
```diff
@@ -0,0 +1,98 @@
+
+    public MCRBasicFileAttributesProxy(P path) {
+        this.path = path;
+    }
```
This is not how attributes work. Call to the methods should not result in exceptions once they are constructed:
`Files.readAttributes(file, MyFileAttributes.class)` should throw any exception and not a call to lastAccessTime(). Often a check for existence of a file is implemented as reading the file attributes.

Mewel replied:
Ok, will rm this class entirely and solve the attribute loading in ocfl

yagee-de commented on mycore-base/src/main/java/org/mycore/frontend/fileupload/MCRUploadHandlerIFS.java:
```diff
@@ -199,7 +199,7 @@ public boolean acceptFile(String path, String checksum, long length) throws Exce
                 MCRFileAttributes attrs = Files.readAttributes(child, MCRFileAttributes.class);
                 shouldAcceptFile = attrs.size() != length
-                    || !(checksum.equals(attrs.md5sum()) && child.getFileSystem().verifies(child, attrs));
+                    || !(checksum.equals(attrs.digest()) && child.getFileSystem().verifies(child, attrs));
```
String.equals(MCRDigest) should always be false

Mewel committed:
MCR-3096 some changes

-uppercase MD5Algorithm
-rm AttributeProxy
-fix checksum string digest compare
-inline ALLOWED_ATTRIBUTES
